domain/transfer/receiver/service.go

```diff
@@ -41,12 +45,106 @@ func NewTransferService(authService auth.Service, localFHIRClientFactory fhir.Fa
 	}
 }
 
-func (s service) CreateOrUpdate(ctx context.Context, status string, customerID int, senderDID, fhirTaskID string) error {
-	_, err := s.transferRepo.CreateOrUpdate(ctx, status, fhirTaskID, customerID, senderDID)
+func verifyStateUpdate(from, to string) error {
+	// Contains all possible status-updates from: https://informatiestandaarden.nictiz.nl/wiki/vpk:V4.0_FHIR_eOverdracht#Using_Task_to_manage_the_workflow
+	allowed := []struct {
+		from string
+		to   string
+	}{
+		{transfer.RequestedState, transfer.AcceptedState},
+		{transfer.RequestedState, transfer.RejectedState},
+		{transfer.RequestedState, transfer.OnHoldState},
+		{transfer.OnHoldState, transfer.RequestedState},
+		{transfer.OnHoldState, transfer.CancelledState},
+		{transfer.AcceptedState, transfer.InProgressState},
+		{transfer.InProgressState, transfer.CompletedState},
+	}
+
+	for _, update := range allowed {
+		if from == update.from && to == update.to {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("invalid state change from %s to %s", from, to)
+}
+
+func (s service) completeTransfer(ctx context.Context, incomingTransfer *domain.IncomingTransfer, customerDID, senderDID string) error {
+	taskPath := fmt.Sprintf("Task/%s", incomingTransfer.FhirTaskID)
+	client, err := s.getRemoteFHIRClient(ctx, senderDID, customerDID, taskPath)
+	if err != nil {
+		return err
+	}
+
+	// First, get the task
+	task, err := s.getRemoteTransferTask(ctx, client, incomingTransfer.FhirTaskID)
 	if err != nil {
 		return err
 	}
 
+	// Then, get the composition reference
+	var compositionRef string
+
+	for _, input := range task.Input {
+		if *input.Type.Coding[0].Code == fhir.SnomedNursingHandoffCode {
+			compositionRef = fhir.FromStringPtr(input.ValueReference.Reference)
+			break
+		}
+	}
+
+	if compositionRef == "" {
+		return errors.New("unable to find nursing handoff input")
+	}
+
+	// Read the subject from the composition
+	composition := &eoverdracht.Composition{}
+
+	if err := client().ReadOne(ctx, fmt.Sprintf("/%s", compositionRef), composition); err != nil {
+		return fmt.Errorf("failed to read composition: %w", err)
+	}
+
+	// And at last, retrieve the patient and create it on the local FHIR server
+	patient := &resources.Patient{}
+
+	if err := client().ReadOne(ctx, fmt.Sprintf("/%s", fhir.FromStringPtr(composition.Subject.Reference)), patient); err != nil {
+		return fmt.Errorf("failed to read the patient: %w", err)
+	}
+
+	if err := s.localFHIRClientFactory().CreateOrUpdate(ctx, patient); err != nil {
```

Comment:
I don't think you should use the same identifier as the remote patient.

Reply:
I agree but what would be the alternative?